src/ioctls/mod.rs

```diff
@@ -26,6 +29,100 @@ pub mod vm;
 /// is otherwise a direct mapping to Result.
 pub type Result<T> = std::result::Result<T, errno::Error>;
 
+/// A wrapper around the coalesced MMIO ring page.
+#[derive(Debug)]
+pub(crate) struct KvmCoalescedIoRing {
+    addr: *mut kvm_coalesced_mmio_ring,
+    page_size: usize,
+}
+
+impl KvmCoalescedIoRing {
+    /// Maps the coalesced MMIO ring from the vCPU file descriptor.
+    pub(crate) fn mmap_from_fd<F: AsRawFd>(fd: &F) -> Result<Self> {
+        // SAFETY: We trust the sysconf libc function and we're calling it
+        // with a correct parameter.
+        let page_size = match unsafe { libc::sysconf(libc::_SC_PAGESIZE) } {
+            -1 => return Err(errno::Error::last()),
+            ps => ps as usize,
+        };
+
+        let offset = KVM_COALESCED_MMIO_PAGE_OFFSET * page_size as u32;
+        // SAFETY: KVM guarantees that there is a page at offset
+        // KVM_COALESCED_MMIO_PAGE_OFFSET * PAGE_SIZE if the appropriate
+        // capability is available. If it is not, the call will simply
+        // fail.
+        let addr = unsafe {
+            libc::mmap(
+                null_mut(),
+                page_size,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_SHARED,
+                fd.as_raw_fd(),
+                offset.into(),
+            )
+        };
+        if addr == libc::MAP_FAILED {
+            return Err(errno::Error::last());
+        }
+        Ok(Self {
+            addr: addr.cast::<kvm_coalesced_mmio_ring>(),
```

Comment:
Nit: I think it can infer the type of this cast.
```suggestion
            addr: addr.cast(),
```